cdap-kubernetes/src/main/java/io/cdap/cdap/k8s/runtime/KubeTwillContext.java

```diff
@@ -211,101 +228,304 @@ public String getUID() {
-        return client;
+    @Override
+    public void resourceAdded(DynamicKubernetesObject resource) {
+      resourceModified(resource);
+    }
+
+    @Override
+    public void resourceModified(DynamicKubernetesObject resource) {
+      // For Job type, the number of instances is in the spec.parallelism field, otherwise it is in
+      // the status.replicas field
+      int count;
+      if ("Job".equals(ownerRef.getKind())) {
+        JsonObject spec = resource.getRaw().getAsJsonObject("spec");
+        if (spec == null) {
+          return;
+        }
+        count = spec.getAsJsonPrimitive("parallelism").getAsInt();
+      } else {
+        JsonObject status = resource.getRaw().getAsJsonObject("status");
+        if (status == null) {
+          return;
+        }
+        JsonPrimitive value = status.getAsJsonPrimitive("replicas");
+        if (value == null) {
+          return;
+        }
+        count = value.getAsInt();
       }
 
-      @Override
-      public void resourceAdded(Object resource) {
-        resourceModified(resource);
+      int oldInstances = instances.getAndSet(count);
+      if (oldInstances != count) {
+        LOG.debug("Number of instances change to {}", count);
+        executor.execute(() -> assignInstanceId(api));
       }
 
-      @Override
-      public void resourceModified(Object resource) {
-        // The CustomObjectsApi returns Map<String, Map> to represent the nested structure of the resource
-        Object statusReplicas = getValue(getValue(resource, "status"), "replicas");
-        if (statusReplicas == null) {
-          return;
+      instancesFuture.complete(instances);
+    }
+
+    /**
+     * Assigns the instance Id for this process.
+     */
+    private void assignInstanceId(DynamicKubernetesApi api) {
+      // For StatefulSet, the instance ID is the suffix number.
+      if ("StatefulSet".equals(ownerRef.getKind())) {
+        int idx = podInfo.getName().lastIndexOf('-');
+        try {
+          if (idx >= 0) {
+            setInstanceId(Integer.parseInt(podInfo.getName().substring(idx + 1)));
+            return;
+          }
+        } catch (NumberFormatException e) {
+          LOG.warn("Failed to parse instance id from pod name {}", podInfo.getName(), e);
         }
-        if (!(statusReplicas instanceof Number)) {
+      }
+
+      // If it is not StatefulSet or failed to extract instanceId from the pod name, use the assignment logic
+      performAssignInstanceId(api);
+    }
+
+    /**
+     * Performs the instance Id assignment logic using annotations on the owner object.
+     * The logic involves setting an annotation with key {@link #INSTANCES_ASSIGNMENT}, and value is a Json map
+     * that stores the associate of instance Id to pod name ({@code instanceId -> Pod Name}).
+     * Each of the pod instances will race to update the assignment map, which we rely on the compare and set property
+     * of the k8s replace API to guarantee a consistent view across all pods.
```

Comment:
should we say `put API` instead?

Reply:
The http method is PUT, but the k8s api call it Replace. See https://v1-18.docs.kubernetes.io/docs/reference/generated/kubernetes-api/v1.18/#-strong-write-operations-replicaset-v1-apps-strong-